lib/cocaine/command_line/multi_pipe.rb

```diff
@@ -29,19 +29,43 @@ def close_write
       end
 
       def read
-        @stdout_output = read_stream(@stdout_in)
-        @stderr_output = read_stream(@stderr_in)
+        read_streams(@stdout_in, @stderr_in)
       end
 
       def close_read
-        @stdout_in.close
-        @stderr_in.close
+        @stdout_in.close rescue IOError
+        @stderr_in.close rescue IOError
+      end
+
+      def read_streams(output, error)
+        @stdout_output = ""
+        @stderr_output = ""
+        read_fds = [output,error]
+        while read_fds.size > 0
+          to_read, _ = IO.select(read_fds)
+          if to_read.include?(output)
+            @stdout_output << read_stream(output)
+            read_fds.delete(output) if output.closed?
+          end
+
+          if to_read.include?(error)
+            @stderr_output << read_stream(error)
+            read_fds.delete(error) if error.closed?
+          end
+        end
+
       end
 
       def read_stream(io)
         result = ""
-        while partial_result = io.read(8192)
-          result << partial_result
+        begin
+          while partial_result = io.read_nonblock(8192)
+            result << partial_result
+          end
+        rescue EOFError, Errno::EPIPE
```

Comment:
Hound brings up a good point. Don't we want to handle these situations? EPIPE and EINTR seem especially important to do something with. I can see ignoring EWOULDBLOCK and EAGAIN though.

Reply:
EPIPE means the pipe closed, so we can't read any more. We could check exit status, but that happens elsewhere, right? Here, we just know that because the pipe is gone we can close the IO and continue on.

EWOULDBLOCK, EINTR and EAGAIN just have us continue and retry. I don't think we want to do anything special on EINTR. Would you do something?